docs/services/sdk/side-effects.mdx

```diff
@@ -209,6 +218,86 @@ You can throw a terminal exception via `dev.restate.sdk.core.TerminalException`
 </TabItem>
 </Tabs>
 
+## Promise/Awaitable combinators
+
+<Tabs groupId="sdk" queryString>
+<TabItem value="ts" label="TypeScript">
+The SDK provides combinators for working with promises.
+
+**CombineablePromise.all():**
+
+```shell
+const result = await CombineablePromise.all([promise1, promise2]);
+```
+Creates a Promise that is either:
+1. Resolved with an array of results, when all of the provided Promises resolve.
+2. Rejected when any Promise is rejected.
+
+Similar to [`Promise.all()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all),
+but the outcome is stored in the Restate journal to be deterministically replayable.
+
+**CombineablePromise.any():**
+
+```shell
+const result = await CombineablePromise.any([promise1, promise2]);
+```
+
+Creates a promise with two possible outcomes:
+1. The promise resolves with the first successful result from the input promises, once any of them resolves.
+2. The promise gets rejected when all the input promises are rejected (including when an empty iterable is passed),
+returning an `AggregateError` containing an array of the reasons for rejection.
+
+Similar to [`Promise.any()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any),
+but the outcome is stored in the Restate journal to be deterministically replayable.
+
+**CombineablePromise.race():**
+
+```shell
+const result = await CombineablePromise.race([promise1, promise2]);
+```
+
+Creates a Promise that is:
+1. Resolved when any of the provided Promises are resolved.
+2. Rejected when any of the provided Promises are rejected.
+
+Similar to [`Promise.race()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race),
+but the outcome is stored in the Restate journal to be deterministically replayable.
+
+**CombineablePromise.allSettled():**
+
+```shell
+const result = await CombineablePromise.allSettled([promise1, promise2]);
+```
+
+Creates a promise that resolves once all the input promises have settled (including when an empty iterable is passed).
+It returns an array of objects describing the outcome of each promise, whether resolved or rejected.
+
+Similar to [`Promise.allSettled()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled),
+but the outcome is stored in the Restate journal to be deterministically replayable.
+
+</TabItem>
+<TabItem value="java" label="Java">
+
+The SDK provides combinators for working with `Awaitable`.
+
+**Awaitable.all():**
+
+```shell
+SomeResponse res = (SomeResponse) Awaitable.all(a1, a2, a3).await();
+```
+
+Creates an Awaitable that awaits all of the provided Awaitables to resolve.
+
+**Awaitable.any():**
+
+```shell
+SomeResponse res = (SomeResponse) Awaitable.any(a1, a2, a3).await();
+```
+
+Creates an Awaitable that awaits any of the provided Awaitables to resolve.
```

Comment:
You can write down the the behaviour is similar to JDK's `CompletableFuture.all/any`